### PR TITLE
[SPARK-45265][SQL] Support Hive 4.0 metastore

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR continues the work from #43064 and #45801 to support Hive Metastore Server 4.0. CHAR/VARCHAR type partition filter pushdown is not included in this PR, as it requires further investment.

### Why are the changes needed?

Enhance the multiple hive metastore server support feature

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?

Passing HiveClient*Suites w/ 4.0

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
👍🏻  Thank you so much for taking over this, @yaooqinn .

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -1664,7 +1664,7 @@ object TestSettings {
         "--add-opens=java.base/sun.util.calendar=ALL-UNNAMED",
         "-Djdk.reflect.useDirectMethodHandle=false",
         "-Dio.netty.tryReflectionSetAccessible=true").mkString(" ")
-      s"-Xmx$heapSize -Xss4m -XX:MaxMetaspaceSize=$metaspaceSize -XX:ReservedCodeCacheSize=128m -Dfile.encoding=UTF-8 $extraTestJavaArgs"
+      s"-Xmx$heapSize -Xss64m -XX:MaxMetaspaceSize=$metaspaceSize -XX:ReservedCodeCacheSize=128m -Dfile.encoding=UTF-8 $extraTestJavaArgs"
```
Comment (yaooqinn) [inline project/SparkBuild.scala:1667]: A super long generated query in HivePartitionFilteringSuite causes derby stackoverflow

**Comment by yaooqinn** [comment]
All GitHub actions are passing now. Please take a look. @LuciferYang @dongjoon-hyun @HyukjinKwon @cloud-fan

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -1183,6 +1183,14 @@ object Hive {
...
     (Test / javaOptions) := (Test / javaOptions).value.filterNot(_.contains("MaxMetaspaceSize")),
     (Test / javaOptions) += "-XX:MaxMetaspaceSize=2g",
+    // SPARK-45265: HivePartitionFilteringSuite addPartitions related tests generate supper long
+    // direct sql against derby server, which may cause stack overflow error when derby do sql
+    // parsing.
+    // We need to increase the Xss for the test. Meanwhile, QueryParsingErrorsSuite requires a
+    // smaller size of Xss to mock a FAILED_TO_PARSE_TOO_COMPLEX error, so we need to set for
+    // hive moudle specifically.
+    (Test / javaOptions) := (Test / javaOptions).value.filterNot(_.contains("Xss")),
+    (Test / javaOptions) += "-Xss64m",
```
Comment (LuciferYang) [inline project/SparkBuild.scala:1193]: Subsequent options will override previous ones, so there is no need to filter out the existing -Xss, but that's also fine with me.

Comment (yaooqinn) [reply project/SparkBuild.scala:1193]: just copied this for the metaspace logic

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @yaooqinn .

I believe we can merge this and test more before Apache Spark 4.0.0 release.

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun @LuciferYang

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -177,8 +179,10 @@ private[hive] class HiveClientImpl(
     // got changed. We reset it to clientLoader.ClassLoader here.
     state.getConf.setClassLoader(clientLoader.classLoader)
     shim.setCurrentSessionState(state)
-    state.out = new PrintStream(outputBuffer, true, UTF_8.name())
-    state.err = new PrintStream(outputBuffer, true, UTF_8.name())
+    val clz = state.getClass.getField("out").getType.asInstanceOf[Class[_ <: PrintStream]]
+    val ctor = clz.getConstructor(classOf[OutputStream], classOf[Boolean], classOf[String])
+    state.getClass.getField("out").set(state, ctor.newInstance(outputBuffer, true, UTF_8.name()))
```
Comment (hidataplus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:184]: build fail ： ”the result type of an implicit conversion must be more specific than Object“

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 4]
```diff
@@ -1030,7 +1030,7 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     }
     val metaStoreParts = partsWithLocation
       .map(p => p.copy(spec = toMetaStorePartitionSpec(p.spec)))
-    client.createPartitions(db, table, metaStoreParts, ignoreIfExists)
+    client.createPartitions(tableMeta, metaStoreParts, ignoreIfExists)
```
Comment (hidataplus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1033]: why?

**Pull request closed without merging**